### DISCUSSION TITLE
A tutorial would be nice

### DISCUSSION THREAD

**Discussion by gergelypolonkai** [beat-feet/beat-feet#72] [Ideas]
I just downloaded this game from F-Droid, and itʼs really great!

My only problem is that i either experience a serious sound lag, or i donʼt get the basics of the game (it can safely be any of the two).

To make sure itʼs not the latter, you might want to introduce a tutorial. This might involve a simple, monoton drum beat to which you should jump.

To make sure itʼs not the former, you might introduce a new settings screen that plays the aforementioned drum beat to which you need to tap on the screen.This way you can detect the lag between the sound output, the screen, and even the playerʼs reflexes. This would also make the game much more accessible.

Also, an extra feature might come in handy for people with less sense for rhythm: rhythm bars! The bottom might show little bars that follow the musicʼs rhythm you are supposed to follow.

I remember seeing some of these features in other rhythm based games, probably in Beat Buddy or Necrodancer. They can help a lot.

**Comment by pserwylo** [comment]
Thanks for the feedback @gergelypolonkai. I would love to show this sort of information, and indeed I like that there is the black space a tthe bottom of the level we can use to show some more feedback about the music playing. Unfortunately I still don't have a good idea of what the actual *beat* of a given song is, beacuse I lack certain skills in digital signal processing (see #43). I could manually do it, but I'd much prefer to have a generic approach. 

Regarding lag, could this be the 3 second wait I've put in at the start of each level which is silence? I added that because it was really frustrating to be dumped straight into the level.

If so, then maybe there could be one common 3 second sound we play at the start of each level leading up to the beginning of the sound to avoid that awkward silence?

**Reply by gergelypolonkai** [reply]
The problem is definitely not the first three seconds. It feels like i have to jump to a different beat than the music has. Itʼs either because

a) i donʼt understand how the game processes music and where it think i need to jump; thus, the understanding of the song by the game and by me is different (thatʼs why i thought the visual feedback would be nice, to avoid this case), or
b) the music plays with a few hundred milliseconds behind/ahead the level itself. So when there is an obstacle, the corresponding music beat is already happened, or will happen slightly in the future. Visual feedback doesnʼt help in this case, as the problem is that the visuals and the sound is out of sync. Thatʼs why i suggested to add a sync mini-game, so the game knows how much time before/after the visuals should it play the music.

**Reply by pserwylo** [reply]
Cool (not about the lag, but about the prompt and detailed response!).

There are three likely things that are happening:
* Genuine lag, due to poor resource management, excessive calculations, etc, etc. Proper profiling of the code will be able to address this by increasing the games performance and responsiveness.
* Obstacles not being offset with the player. An obstacle in the game corresponds to a feature in the music, not the time at which the player is expected to jump. I've experimented in earlier (ha! the whole thing is still pretty early!) with offsetting the obstacles by a fraction of a second, so that the *time at which you need to press jump* lines up with the feature in the music*, but it was hard to get right so I left as is for now.
* The music-feature-extraction is just not very good (see #43 and #74). This is the number one thing I'd love to address, but alas my digital-signal-processing skills are not particularly good, so it is all a bit of trial and error right now. 

I will post a video of me playing later in the week so that we can see whether you're experiencing the game as I experience it, or with a distinctively different sort of lag.

**Reply by gergelypolonkai** [reply]
I guess it might even be the combination of the three.

1. Genuine lag. My phone is no longer high-end (a Huawei P20 Pro), but pretty strong; however, i run a huge lot of applications that might hog the CPU and RAM. I’ll try it on a less “crowded” phone as soon as i find it :grinning:
2. Obstacles not being offset with the player. Having a built-in offset between the musing and the level would be *great*, because (IMHO) most players will tap for the rhythm of the music, not a few milliseconds sooner. Also, the warm up/lag calculation mini-game would solve exactly this problem.
3. music feature extraction problems. If your signal processing skills are weak, mine are completely non-existing. It may or may not be the source of my problem. However, i noticed several times that an obstacle is coming at my way that i can’t jump across (e.g. a really tall lamp or building in the middle of nowhere; probably doable with a double-jump, but most probably *not* at the rhythm of the music).
